File: backend/app/schemas/db_views.py
"""新版资料库视图 Schema（只读）。"""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, field_validator


class GroupMemberRow(BaseModel):
    membership_id: int
    artist_id: int
    artist_uid: Optional[str] = None
    name: str
    stage_name: Optional[str] = None
    avatar_path: Optional[str] = None
    join_date: Optional[str] = None
    leave_date: Optional[str] = None
    status: str
    positions: List[str] = []
    is_dangling: bool = False
    chinese_name: Optional[str] = None
    english_name: Optional[str] = None
    korean_name: Optional[str] = None
    birth_date: Optional[str] = None
    birth_place: Optional[str] = None
    gender: Optional[str] = None
    debut_date: Optional[str] = None
    occupation: Optional[str] = None
    description: Optional[str] = None
    tagline: Optional[str] = None
    social_media: Optional[dict] = None
    external_links: List[dict] = []


class GroupCompanyRow(BaseModel):
    relation_id: int
    company_id: int
    name: str
    role: Optional[str] = None
    status: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    is_dangling: bool = False


class GroupIssueCount(BaseModel):
    error: int
    warning: int
    hint: int


class GroupListRow(BaseModel):
    id: int
    uid: str
    name: str
    korean_name: Optional[str] = None
    chinese_name: Optional[str] = None
    debut_date: Optional[str] = None
    group_type: Optional[str] = None
    avatar_path: Optional[str] = None
    is_subunit: bool = False
    parent_name: Optional[str] = None
    members_active: int
    members_former: int
    member_sample: List[dict] = []
    members: List[GroupMemberRow] = []
    companies: List[GroupCompanyRow] = []
    works: dict
    completeness: int
    completeness_missing: List[str] = []
    issues: GroupIssueCount


class GroupListResponse(BaseModel):
    items: List[GroupListRow]
    total: int
    checklist: List[str]


class GroupWorkspaceGroup(BaseModel):
    id: int
    uid: str
    name: str
    chinese_name: Optional[str] = None
    english_name: Optional[str] = None
    korean_name: Optional[str] = None
    debut_date: Optional[str] = None
    group_type: Optional[str] = None
    gender_type: Optional[str] = None
    origin_country: Optional[str] = None
    description: Optional[str] = None
    tagline: Optional[str] = None
    avatar_path: Optional[str] = None
    banner_path: Optional[str] = None
    social_media: Optional[dict] = None
    hide_from_home: bool = False
    external_links: List[dict] = []


class WorkspaceSubUnit(BaseModel):
    uid: str
    id: int
    name: str
    group_type: Optional[str] = None
    completeness: int = 0
    members_active: int = 0
    member_ids: List[int] = []


class WorkspaceAlbum(BaseModel):
    id: int
    uid: str
    name: str
    release_date: Optional[str] = None
    album_type: Optional[str] = None
    track_count: int = 0
    video_count: int = 0


class GroupWorkspaceResponse(BaseModel):
    group: GroupWorkspaceGroup
    row: Optional[GroupListRow] = None
    sub_units: List[WorkspaceSubUnit] = []
    albums: List[WorkspaceAlbum] = []
    issues: List[dict] = []
    issue_counts: GroupIssueCount
    completeness_checklist: List[str] = []


class SourcePreviewRequest(BaseModel):
    urls: List[str] = []
    url: Optional[str] = None  # 兼容单链接写法
    step: Optional[str] = None  # 1-6 或 identity/subunits/members/member_details/albums/tracks；省略=全量
    targets: Optional[List[str]] = None  # 搜索补全：多选目标，优先于 step

    @field_validator("step", mode="before")
    @classmethod
    def _coerce_step(cls, v):
        if v is None or v == "":
            return None
        return str(v)


class SourceDiffItem(BaseModel):
    field: str
    label: str
    current: Optional[str] = None
    proposed: str
    sources: List[str] = []
    source_urls: List[str] = []
    conflict: bool = False
    is_new: bool = False
    same: bool = False


class SourceMemberProposal(BaseModel):
    action: str  # create / link / exists / enrich
    name: str
    korean_name: Optional[str] = None
    join_date: Optional[str] = None
    leave_date: Optional[str] = None
    status: str
    sources: List[str] = []
    artist_id: Optional[int] = None
    artist_uid: Optional[str] = None
    positions: Optional[List[str]] = None
    artist_fields: dict = {}
    # 字段来源：ai | fandom | wikidata | wikipedia | baidu | crawl
    field_origins: dict = {}
    ai_basis: Optional[str] = None
    possible_duplicate_of: List[str] = []


class SourceCompanyProposal(BaseModel):
    action: str  # create / link / exists
    name: str
    company_id: Optional[int] = None
    role: Optional[str] = None
    sources: List[str] = []


class SourceSubunitProposal(BaseModel):
    action: str  # create / exists
    name: str
    member_names: List[str] = []
    sources: List[str] = []


class SourceAlbumProposal(BaseModel):
    action: str  # create / exists / fill_tracks / update
    name: str
    year: Optional[str] = None

    @field_validator("year", mode="before")
    @classmethod
    def _coerce_year(cls, v):
        if v is None or v == "":
            return None
        return str(v)
    release_date: Optional[str] = None
    track_count: Optional[int] = None
    record_type: Optional[str] = None
    extractor: Optional[str] = None
    external_id: Optional[str] = None
    cover_url: Optional[str] = None
    sources: List[str] = []
    album_id: Optional[int] = None
    tracks_preview: List[dict] = []
    db_release_date: Optional[str] = None
    patch: dict = {}
    candidates: List[dict] = []
    needs_tracks: bool = False


class SourcePreviewResponse(BaseModel):
    sources: List[dict] = []
    items: List[SourceDiffItem]
    member_proposals: List[SourceMemberProposal] = []
    subunit_proposals: List[SourceSubunitProposal] = []
    company_proposals: List[SourceCompanyProposal] = []
    album_proposals: List[SourceAlbumProposal] = []
    single_proposals: List[SourceAlbumProposal] = []
    album_note: Optional[str] = None
    extra: dict = {}
    errors: List[str] = []


class TracklistPreviewRequest(BaseModel):
    album_id: int
    external_id: str


class TracklistPreviewResponse(BaseModel):
    album_id: int
    album_name: str
    external_id: str
    release_date: Optional[str] = None
    tracks_preview: List[dict] = []


class SourceApplyRequest(BaseModel):
    fields: dict = {}
    members: List[dict] = []
    albums: List[dict] = []
    subunits: List[dict] = []
    companies: List[dict] = []
    source_urls: List[str] = []
    step: Optional[str] = None  # 与 preview 同；省略=全量（向后兼容）

    @field_validator("step", mode="before")
    @classmethod
    def _coerce_step(cls, v):
        if v is None or v == "":
            return None
        return str(v)


class SourceApplyResponse(BaseModel):
    applied: dict


class SourceCandidate(BaseModel):
    source_type: str
    title: str
    snippet: str = ""
    url: str


class SourceSearchResponse(BaseModel):
    items: List[SourceCandidate]


class ArtistMembershipRow(BaseModel):
    membership_id: int
    group_id: int
    group_uid: str
    group_name: str
    group_type: Optional[str] = None
    join_date: Optional[str] = None
    leave_date: Optional[str] = None
    status: str
    positions: List[str] = []


class ArtistWorkspaceArtist(BaseModel):
    id: int
    uid: str
    name: str
    korean_name: Optional[str] = None
    chinese_name: Optional[str] = None
    english_name: Optional[str] = None
    stage_name: Optional[str] = None
    birth_date: Optional[str] = None
    birth_place: Optional[str] = None
    gender: Optional[str] = None
    debut_date: Optional[str] = None
    occupation: Optional[str] = None
    description: Optional[str] = None
    tagline: Optional[str] = None
    avatar_path: Optional[str] = None
    banner_path: Optional[str] = None
    social_media: Optional[dict] = None
    hide_from_home: bool = False
    external_links: List[dict] = []


class ArtistWorkspaceResponse(BaseModel):
    artist: ArtistWorkspaceArtist
    memberships: List[ArtistMembershipRow] = []
    companies: List[dict] = []
    albums: List[WorkspaceAlbum] = []
    songs: List[WorkspaceArtistSong] = []
    songs_count: int = 0
    videos: List[WorkspaceVideoBrief] = []
    completeness: int = 0
    completeness_missing: List[str] = []
    issues: List[dict] = []
    issue_counts: dict = {}
    completeness_checklist: List[str] = []


# ===== 资料库列表：艺人 / 专辑 / 歌曲（只读，与组合列表同口径）=====


class ArtistGroupBrief(BaseModel):
    group_id: int
    group_uid: str
    name: str
    status: str


class ArtistListRow(BaseModel):
    id: int
    uid: str
    name: str
    chinese_name: Optional[str] = None
    english_name: Optional[str] = None
    korean_name: Optional[str] = None
    stage_name: Optional[str] = None
    birth_date: Optional[str] = None
    debut_date: Optional[str] = None
    gender: Optional[str] = None
    avatar_path: Optional[str] = None
    groups: List[ArtistGroupBrief] = []
    works: dict
    completeness: int
    completeness_missing: List[str] = []
    issues: GroupIssueCount


class ArtistListResponse(BaseModel):
    items: List[ArtistListRow]
    total: int
    checklist: List[str]


class AlbumListRow(BaseModel):
    id: int
    uid: str
    name: str
    chinese_name: Optional[str] = None
    korean_name: Optional[str] = None
    album_type: Optional[str] = None
    release_date: Optional[str] = None
    cover_path: Optional[str] = None
    release_artist_type: Optional[str] = None
    release_artist_name: Optional[str] = None
    release_artist_uid: Optional[str] = None
    works: dict
    completeness: int
    completeness_missing: List[str] = []
    issues: GroupIssueCount


class AlbumListResponse(BaseModel):
    items: List[AlbumListRow]
    total: int
    checklist: List[str]


class SongPerformerBrief(BaseModel):
    subject_type: str
    subject_id: int
    subject_uid: Optional[str] = None
    name: str
    role: str


class SongListRow(BaseModel):
    id: int
    uid: str
    name: str
    chinese_name: Optional[str] = None
    korean_name: Optional[str] = None
    song_type: Optional[str] = None
    release_date: Optional[str] = None
    duration: Optional[int] = None
    performers: List[SongPerformerBrief] = []
    works: dict
    completeness: int
    completeness_missing: List[str] = []
    issues: GroupIssueCount


class SongListResponse(BaseModel):
    items: List[SongListRow]
    total: int
    checklist: List[str]


class AlbumWorkspaceSubject(BaseModel):
    type: str
    id: int
    uid: str
    name: str


class AlbumWorkspaceLabel(BaseModel):
    id: int
    name: str


class AlbumWorkspaceTrack(BaseModel):
    track_id: int
    song_id: Optional[int] = None
    song_uid: Optional[str] = None
    song_name: Optional[str] = None
    song_chinese_name: Optional[str] = None
    song_duration: Optional[int] = None
    disc_number: int = 1
    track_number: int = 0
    is_dangling: bool = False


class AlbumWorkspaceVideo(BaseModel):
    id: int
    uid: str
    name: str
    video_type: Optional[str] = None


class AlbumWorkspaceAlbum(BaseModel):
    id: int
    uid: str
    name: str
    chinese_name: Optional[str] = None
    english_name: Optional[str] = None
    korean_name: Optional[str] = None
    release_date: Optional[str] = None
    album_type: Optional[str] = None
    description: Optional[str] = None
    cover_path: Optional[str] = None


class AlbumWorkspaceResponse(BaseModel):
    album: AlbumWorkspaceAlbum
    subject: Optional[AlbumWorkspaceSubject] = None
    label: Optional[AlbumWorkspaceLabel] = None
    tracks: List[AlbumWorkspaceTrack] = []
    videos: List[AlbumWorkspaceVideo] = []
    completeness: int = 0
    completeness_missing: List[str] = []
    issues: List[dict] = []
    issue_counts: dict = {}
    completeness_checklist: List[str] = []


class SongWorkspacePerformer(BaseModel):
    relation_id: int
    subject_type: str
    subject_id: Optional[int] = None
    subject_uid: Optional[str] = None
    name: str
    role: str
    order: int = 0
    is_dangling: bool = False


class SongWorkspaceAlbum(BaseModel):
    track_id: int
    id: int
    uid: str
    name: str
    release_date: Optional[str] = None
    album_type: Optional[str] = None
    disc_number: int = 1
    track_number: int = 0


class SongWorkspaceVideo(BaseModel):
    id: int
    uid: str
    name: str
    video_type: Optional[str] = None


class SongWorkspaceCredit(BaseModel):
    id: int
    name: str
    role: str
    artist_id: Optional[int] = None
    artist_name: Optional[str] = None


class SongWorkspaceSong(BaseModel):
    id: int
    uid: str
    name: str
    chinese_name: Optional[str] = None
    english_name: Optional[str] = None
    korean_name: Optional[str] = None
    song_type: Optional[str] = None
    release_date: Optional[str] = None
    duration: Optional[int] = None
    description: Optional[str] = None


class SongWorkspaceResponse(BaseModel):
    song: SongWorkspaceSong
    performers: List[SongWorkspacePerformer] = []
    albums: List[SongWorkspaceAlbum] = []
    videos: List[SongWorkspaceVideo] = []
    credits: List[SongWorkspaceCredit] = []
    completeness: int = 0
    completeness_missing: List[str] = []
    issues: List[dict] = []
    issue_counts: dict = {}
    completeness_checklist: List[str] = []


class WorkspaceArtistSong(BaseModel):
    id: int
    uid: str
    name: str
    chinese_name: Optional[str] = None
    release_date: Optional[str] = None
    duration: Optional[int] = None


class WorkspaceVideoBrief(BaseModel):
    id: int
    uid: str
    name: str
    video_type: Optional[str] = None


class FieldSuggestRequest(BaseModel):
    field: str


class FieldSuggestResponse(BaseModel):
    field: str
    value: Optional[str] = None
    source_label: Optional[str] = None
    source_url: Optional[str] = None


class ArtistApplyResult(BaseModel):
    fields: List[str] = []
    fields_skipped_locked: List[str] = []


class ArtistApplyResponse(BaseModel):
    applied: ArtistApplyResult

File: frontend/src/api/dbViews.ts
import http from './index'

export interface DbMemberRow {
  membership_id: number
  artist_id: number
  artist_uid?: string | null
  name: string
  stage_name?: string | null
  avatar_path?: string | null
  join_date?: string | null
  leave_date?: string | null
  status: string
  positions: string[]
  is_dangling: boolean
  chinese_name?: string | null
  english_name?: string | null
  korean_name?: string | null
  birth_date?: string | null
  birth_place?: string | null
  gender?: string | null
  debut_date?: string | null
  occupation?: string | null
  description?: string | null
  tagline?: string | null
  social_media?: Record<string, string> | null
  external_links: { url?: string; source?: string; [k: string]: unknown }[]
}

export interface DbCompanyRow {
  relation_id: number
  company_id: number
  name: string
  role?: string | null
  status: string
  start_date?: string | null
  end_date?: string | null
  is_dangling: boolean
}

export interface DbGroupRow {
  id: number
  uid: string
  name: string
  korean_name?: string | null
  chinese_name?: string | null
  debut_date?: string | null
  group_type?: string | null
  avatar_path?: string | null
  is_subunit: boolean
  parent_name?: string | null
  members_active: number
  members_former: number
  member_sample: { id: number; name: string; avatar_path?: string | null }[]
  members: DbMemberRow[]
  companies: DbCompanyRow[]
  works: { albums: number; songs: number; videos: number }
  completeness: number
  completeness_missing: string[]
  issues: { error: number; warning: number; hint: number }
}

export interface GroupListResponse {
  items: DbGroupRow[]
  total: number
  checklist: string[]
}

export interface DbIssueCount {
  error: number
  warning: number
  hint: number
}

export interface DbArtistGroupBrief {
  group_id: number
  group_uid: string
  name: string
  status: string
}

export interface DbArtistRow {
  id: number
  uid: string
  name: string
  chinese_name?: string | null
  english_name?: string | null
  korean_name?: string | null
  stage_name?: string | null
  birth_date?: string | null
  debut_date?: string | null
  gender?: string | null
  avatar_path?: string | null
  groups: DbArtistGroupBrief[]
  works: { albums: number; songs: number; videos: number }
  completeness: number
  completeness_missing: string[]
  issues: DbIssueCount
}

export interface ArtistListResponse {
  items: DbArtistRow[]
  total: number
  checklist: string[]
}

export interface DbAlbumRow {
  id: number
  uid: string
  name: string
  chinese_name?: string | null
  korean_name?: string | null
  album_type?: string | null
  release_date?: string | null
  cover_path?: string | null
  release_artist_type?: string | null
  release_artist_name?: string | null
  release_artist_uid?: string | null
  works: { tracks: number; videos: number }
  completeness: number
  completeness_missing: string[]
  issues: DbIssueCount
}

export interface AlbumListResponse {
  items: DbAlbumRow[]
  total: number
  checklist: string[]
}

export interface DbSongPerformer {
  subject_type: string
  subject_id: number
  subject_uid?: string | null
  name: string
  role: string
}

export interface DbSongRow {
  id: number
  uid: string
  name: string
  chinese_name?: string | null
  korean_name?: string | null
  song_type?: string | null
  release_date?: string | null
  duration?: number | null
  performers: DbSongPerformer[]
  works: { albums: number; videos: number }
  completeness: number
  completeness_missing: string[]
  issues: DbIssueCount
}

export interface SongListResponse {
  items: DbSongRow[]
  total: number
  checklist: string[]
}

export interface AlbumWorkspaceSubject {
  type: 'group' | 'artist' | string
  id: number
  uid: string
  name: string
}

export interface AlbumWorkspaceLabel {
  id: number
  name: string
}

export interface AlbumWorkspaceTrack {
  track_id: number
  song_id?: number | null
  song_uid?: string | null
  song_name?: string | null
  song_chinese_name?: string | null
  song_duration?: number | null
  disc_number: number
  track_number: number
  is_dangling: boolean
}

export interface AlbumWorkspaceVideo {
  id: number
  uid: string
  name: string
  video_type?: string | null
}

export interface AlbumWorkspaceResponse {
  album: {
    id: number
    uid: string
    name: string
    chinese_name?: string | null
    english_name?: string | null
    korean_name?: string | null
    release_date?: string | null
    album_type?: string | null
    description?: string | null
    cover_path?: string | null
  }
  subject: AlbumWorkspaceSubject | null
  label: AlbumWorkspaceLabel | null
  tracks: AlbumWorkspaceTrack[]
  videos: AlbumWorkspaceVideo[]
  completeness: number
  completeness_missing: string[]
  issues: {
    severity: 'error' | 'warning' | 'hint'
    title: string
    suggestion: string
  }[]
  issue_counts: Record<string, number>
  completeness_checklist: string[]
}

export const albumWorkspaceApi = {
  getAlbumWorkspace(uid: string) {
    return http.get<AlbumWorkspaceResponse>(`/db/albums/${uid}/workspace`).then((r) => r.data)
  },
}

export interface SongWorkspaceResponse {
  song: {
    id: number
    uid: string
    name: string
    chinese_name?: string | null
    english_name?: string | null
    korean_name?: string | null
    song_type?: string | null
    release_date?: string | null
    duration?: number | null
    description?: string | null
  }
  performers: {
    relation_id: number
    subject_type: 'artist' | 'group' | 'dangling' | string
    subject_id?: number | null
    subject_uid?: string | null
    name: string
    role: string
    order: number
    is_dangling: boolean
  }[]
  albums: {
    track_id: number
    id: number
    uid: string
    name: string
    release_date?: string | null
    album_type?: string | null
    disc_number: number
    track_number: number
  }[]
  videos: {
    id: number
    uid: string
    name: string
    video_type?: string | null
  }[]
  credits: {
    id: number
    name: string
    role: string
    artist_id?: number | null
    artist_name?: string | null
  }[]
  completeness: number
  completeness_missing: string[]
  issues: {
    severity: 'error' | 'warning' | 'hint'
    title: string
    suggestion: string
  }[]
  issue_counts: Record<string, number>
  completeness_checklist: string[]
}

export const songWorkspaceApi = {
  getSongWorkspace(uid: string) {
    return http.get<SongWorkspaceResponse>(`/db/songs/${uid}/workspace`).then((r) => r.data)
  },
}

export interface FieldSuggestResult {
  field: string
  value: string | null
  source_label: string | null
  source_url: string | null
}

export const dbFieldSuggestApi = {
  /** 单字段搜索：只检索建议值，不写库；确认后由前端走常规保存 */
  suggestGroupField(uid: string, field: string) {
    return http.post<FieldSuggestResult>(`/db/groups/${uid}/suggest-field`, { field }).then((r) => r.data)
  },
  suggestArtistField(uid: string, field: string) {
    return http.post<FieldSuggestResult>(`/db/artists/${uid}/suggest-field`, { field }).then((r) => r.data)
  },
}

export interface ArtistSourceDiffItem {
  field: string
  label: string
  current?: string | null
  proposed: string
  sources: string[]
  source_urls: string[]
  conflict: boolean
  is_new: boolean
  same?: boolean
}

export interface ArtistSourcePreview {
  sources: { source_type: string; label: string; url: string }[]
  items: ArtistSourceDiffItem[]
  extra: Record<string, unknown>
  errors: string[]
}

export interface ArtistApplyResult {
  fields: string[]
  fields_skipped_locked: string[]
}

export const artistGrowthApi = {
  /** 艺人生长预览：urls 为空时按名称自动检索 Fandom/维基/TMDB */
  preview(uid: string, urls: string[]) {
    return http
      .post<ArtistSourcePreview>(`/db/artists/${uid}/source/preview`, { urls }, { timeout: 180_000 })
      .then((r) => r.data)
  },
  apply(uid: string, fields: Record<string, string>, sourceUrls: string[]) {
    return http
      .post<{ applied: ArtistApplyResult }>(
        `/db/artists/${uid}/source/apply`,
        { fields, source_urls: sourceUrls },
        { timeout: 120_000 },
      )
      .then((r) => r.data)
  },
  /** solo 专辑生长：iTunes / Deezer / MusicBrainz 聚合检索 */
  suggestAlbums(uid: string) {
    return http
      .post<{ albums: AlbumProposal[]; note: string | null }>(
        `/db/artists/${uid}/albums/suggest`,
        {},
        { timeout: 180_000 },
      )
      .then((r) => r.data)
  },
  applyAlbums(uid: string, albums: AlbumProposal[]) {
    return http
      .post<{
        albums_created: number
        albums_updated: number
        albums_skipped: number
        songs_created: number
        tracks_created: number
        tracklist_failed: number
      }>(
        `/db/artists/${uid}/albums/apply`,
        { albums, include_tracks: true },
        { timeout: 180_000 },
      )
      .then((r) => r.data)
  },
}

export interface AlbumProposal {
  action: 'create' | 'update' | 'exists' | 'fill_tracks' | string
  name: string
  year?: string | null
  release_date?: string | null
  record_type?: string | null
  track_count?: number | null
  external_id?: string | null
  cover_url?: string | null
  sources: string[]
  album_id?: number | null
  patch?: Record<string, string>
}

export interface GroupWorkspaceResponse {
  group: {
    id: number
    uid: string
    name: string
    chinese_name?: string | null
    english_name?: string | null
    korean_name?: string | null
    debut_date?: string | null
    group_type?: string | null
    gender_type?: string | null
    origin_country?: string | null
    description?: string | null
    tagline?: string | null
    avatar_path?: string | null
    banner_path?: string | null
    social_media?: Record<string, string> | null
    hide_from_home?: boolean
    external_links: { url?: string; source?: string; [k: string]: unknown }[]
  }
  row: DbGroupRow | null
  sub_units: {
    uid: string
    id: number
    name: string
    group_type?: string | null
    completeness: number
    members_active: number
    member_ids: number[]
  }[]
  albums: {
    id: number
    uid: string
    name: string
    release_date?: string | null
    album_type?: string | null
    track_count: number
    video_count: number
  }[]
  issues: {
    check: string
    severity: 'error' | 'warning' | 'hint'
    entity_type: string
    entity_id: number
    title: string
    detail: string
    suggestion: string
    deep_link?: string | null
  }[]
  issue_counts: { error: number; warning: number; hint: number }
  completeness_checklist: string[]
}

export const dbViewsApi = {
  getGroupRows() {
    return http.get<GroupListResponse>('/db/groups').then((r) => r.data)
  },
  getArtistRows() {
    return http.get<ArtistListResponse>('/db/artists').then((r) => r.data)
  },
  getAlbumRows() {
    return http.get<AlbumListResponse>('/db/albums').then((r) => r.data)
  },
  getSongRows() {
    return http.get<SongListResponse>('/db/songs').then((r) => r.data)
  },
  getGroupWorkspace(uid: string) {
    return http.get<GroupWorkspaceResponse>(`/db/groups/${uid}/workspace`).then((r) => r.data)
  },
  searchSourceCandidates(uid: string, q: string) {
    return http
      .get<{ items: { source_type: string; title: string; snippet: string; url: string }[] }>(
        `/db/groups/${uid}/source/search`,
        { params: { q } },
      )
      .then((r) => r.data)
  },
  /** step=4 / member_details 时 urls 可为空：后端按在籍成员搜索 Fandom 补全档案 */
  previewSource(uid: string, urls: string[], step?: string | number, targets?: string[]) {
    return http
      .post<{
        sources: { source_type: string; label: string; url: string }[]
        items: {
          field: string
          label: string
          current?: string | null
          proposed: string
          sources: string[]
          source_urls: string[]
          conflict: boolean
          is_new: boolean
        }[]
        member_proposals: {
          action: 'create' | 'link' | 'exists' | 'enrich'
          name: string
          korean_name?: string | null
          join_date?: string | null
          leave_date?: string | null
          status: string
          sources: string[]
          artist_id?: number | null
          artist_uid?: string | null
          positions?: string[] | null
          artist_fields?: {
            birth_date?: string | null
            chinese_name?: string | null
            english_name?: string | null
            stage_name?: string | null
            birth_place?: string | null
            occupation?: string | null
            description?: string | null
          }
          /** 字段来源：ai | fandom | wikidata | wikipedia | baidu | crawl */
          field_origins?: Record<string, string>
          ai_basis?: string | null
          possible_duplicate_of?: string[]
        }[]
        subunit_proposals: {
          action: 'create' | 'exists'
          name: string
          member_names: string[]
          sources: string[]
        }[]
        company_proposals: {
          action: 'create' | 'link' | 'exists'
          name: string
          company_id?: number | null
          role?: string | null
          sources: string[]
        }[]
        album_proposals: {
          action: 'create' | 'exists' | 'fill_tracks' | 'update'
          name: string
          year?: string | null
          release_date?: string | null
          track_count?: number | null
          external_id?: string | null
          cover_url?: string | null
          sources: string[]
          album_id?: number | null
          db_release_date?: string | null
          patch?: Record<string, unknown>
          tracks_preview?: { name?: string; track_number?: number; disc_number?: number }[]
          candidates?: {
            external_id?: string | null
            name?: string | null
            artist?: string | null
            year?: string | null
            track_count?: number | null
            cover_url?: string | null
            source?: string | null
          }[]
          needs_tracks?: boolean
        }[]
        extra: {
          members?: { name?: string; join_date?: string | null; leave_date?: string | null }[]
          company_raw?: string
          fandom_name?: string
          biography?: string
          thumb?: string
          description?: string
          step?: string | null
          baidu_note?: string | null
          ai_gap_fill?: {
            status?: 'skipped' | 'ok' | 'error' | string
            filled_fields?: number
            members_touched?: number
            error?: string | null
          } | null
          [k: string]: unknown
        }
        errors: string[]
      }>(
        `/db/groups/${uid}/source/preview`,
        {
          urls,
          step: step != null ? String(step) : undefined,
          targets: targets?.length ? targets : undefined,
        },
        {
          timeout:
            String(step) === '4' || String(step) === 'member_details' ? 300_000 : 180_000,
        },
      )
      .then((r) => r.data)
  },
  previewTracklist(uid: string, albumId: number, externalId: string) {
    return http
      .post<{
        album_id: number
        album_name: string
        external_id: string
        release_date?: string | null
        tracks_preview: { name?: string; track_number?: number; disc_number?: number }[]
      }>(
        `/db/groups/${uid}/source/tracklist-preview`,
        { album_id: albumId, external_id: externalId },
        { timeout: 120_000 },
      )
      .then((r) => r.data)
  },
  applySource(
    uid: string,
    payload: {
      fields: Record<string, string>
      members: Record<string, unknown>[]
      albums: Record<string, unknown>[]
      subunits: Record<string, unknown>[]
      companies?: Record<string, unknown>[]
      source_urls: string[]
      step?: string | number
    },
  ) {
    const { step, ...rest } = payload
    return http
      .post<{ applied: GrowthApplyResult }>(
        `/db/groups/${uid}/source/apply`,
        { ...rest, step: step != null ? String(step) : undefined },
        { timeout: 180_000 },
      )
      .then((r) => r.data)
  },
}

export interface GrowthApplyResult {
  fields: string[]
  fields_skipped_locked: string[]
  members_created: number
  members_linked: number
  members_skipped: number
  albums_created: number
  albums_updated?: number
  albums_skipped: number
  songs_created: number
  tracks_created: number
  tracklist_failed: number
  subunits_created: number
  subunit_members_created: number
  companies_created?: number
  companies_linked?: number
  companies_skipped?: number
  artist_fields_filled: number
  albums_year_only?: number
  note_year_only_albums?: string
}

export interface ArtistWorkspaceResponse {
  artist: {
    id: number
    uid: string
    name: string
    korean_name?: string | null
    chinese_name?: string | null
    english_name?: string | null
    stage_name?: string | null
    birth_date?: string | null
    birth_place?: string | null
    gender?: string | null
    debut_date?: string | null
    occupation?: string | null
    description?: string | null
    tagline?: string | null
    avatar_path?: string | null
    banner_path?: string | null
    social_media?: Record<string, string> | null
    hide_from_home?: boolean
    external_links: { url?: string; source?: string; [k: string]: unknown }[]
  }
  memberships: {
    membership_id: number
    group_id: number
    group_uid: string
    group_name: string
    group_type?: string | null
    join_date?: string | null
    leave_date?: string | null
    status: string
    positions: string[]
  }[]
  companies: {
    id: number
    company_id: number
    company_name: string
    role?: string | null
    status: string
    start_date?: string | null
    end_date?: string | null
  }[]
  albums: {
    id: number
    uid: string
    name: string
    release_date?: string | null
    album_type?: string | null
    track_count: number
    video_count: number
  }[]
  songs: {
    id: number
    uid: string
    name: string
    chinese_name?: string | null
    release_date?: string | null
    duration?: number | null
  }[]
  songs_count: number
  videos: {
    id: number
    uid: string
    name: string
    video_type?: string | null
  }[]
  completeness: number
  completeness_missing: string[]
  issues: {
    severity: 'error' | 'warning' | 'hint'
    title: string
    suggestion: string
  }[]
  issue_counts: Record<string, number>
  completeness_checklist: string[]
}

export const artistWorkspaceApi = {
  getArtistWorkspace(uid: string) {
    return http.get<ArtistWorkspaceResponse>(`/db/artists/${uid}/workspace`).then((r) => r.data)
  },
}
